Source/Scene/ModelExperimental/MorphTargetsPipelineStage.js

```diff
@@ -0,0 +1,240 @@
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_POSITION =
+  "getMorphedPosition";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_POSITION =
+  "vec3 getMorphedPosition(in vec3 position)";
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_NORMAL = "getMorphedNormal";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_NORMAL =
+  "vec3 getMorphedNormal(in vec3 normal)";
+MorphTargetsPipelineStage.FUNCTION_ID_GET_MORPHED_TANGENT = "getMorphedTangent";
+MorphTargetsPipelineStage.FUNCTION_SIGNATURE_GET_MORPHED_TANGENT =
+  "vec3 getMorphedTangent(in vec3 tangent)";
+
+/**
+ * This pipeline stage processes the morph targets and weights of a primitive,
+ * adding the relevant attribute declarations and functions to the shaders.
+ *
+ * Processes a primitive. This stage modifies the following parts of the render resources:
+ * <ul>
+ *  <li> adds attribute declarations for the morph targets in the vertex shader
+ *  <li> adds the uniform declaration for the morph weights in the vertex shader
+ *  <li> adds functions to apply the morphs in the vertex shader
+ * </ul>
+ *
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} primitive The primitive.
+ *
+ * @private
+ */
+MorphTargetsPipelineStage.process = function (renderResources, primitive) {
+  const shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine(
+    "HAS_MORPH_TARGETS",
+    undefined,
+    ShaderDestination.VERTEX
+  );
+
+  addGetMorphedAttributeFunctionDeclarations(shaderBuilder);
+
+  for (let i = 0; i < primitive.morphTargets.length; i++) {
+    const morphTarget = primitive.morphTargets[i];
+    const attributes = morphTarget.attributes;
+
+    for (let j = 0; j < attributes.length; j++) {
+      const attribute = attributes[j];
+      const semantic = attribute.semantic;
+
+      // Cesium only supports morph targets for positions, normals, and tangents
+      if (
+        semantic !== VertexAttributeSemantic.POSITION &&
+        semantic !== VertexAttributeSemantic.NORMAL &&
+        semantic !== VertexAttributeSemantic.TANGENT
+      ) {
+        continue;
+      }
+
+      processMorphTargetAttribute(
+        renderResources,
+        attribute,
+        renderResources.attributeIndex,
+        i
+      );
+      renderResources.attributeIndex++;
+    }
+  }
+
+  addGetMorphedAttributeFunctionReturns(shaderBuilder);
+
+  const numWeights = primitive.morphWeights.length;
+  shaderBuilder.addUniform(
+    "float",
+    `u_morphWeights[${numWeights}]`,
+    ShaderDestination.VERTEX
+  );
+
+  shaderBuilder.addVertexLines([MorphTargetsStageVS]);
+
+  const uniformMap = {
+    u_morphWeights: function () {
+      return primitive.morphWeights;
+    },
+  };
+
+  renderResources.uniformMap = combine(uniformMap, renderResources.uniformMap);
+};
+
+const scratchAttributeInfo = {};
```

Comment:
I would explicitly initialize this with the two fields explicitly set to undefined. More readable and _might_ make it easier for a JS engine to optimize it as a fixed-size structure.